File: python/src/features/pieces/piece.py
from dataclasses import dataclass


@dataclass
class Piece:
    id: int | None
    title: str
    completed: bool
    created_at: str | None = None

    @staticmethod
    def describe(title: str) -> "Piece":
        if not title or not title.strip():
            raise ValueError("Title must not be empty")
        return Piece(id=None, title=title.strip(), completed=False)

    def toggle(self) -> None:
        self.completed = not self.completed

    def rename(self, title: str) -> None:
        if not title or not title.strip():
            raise ValueError("Title must not be empty")
        self.title = title.strip()

File: typescript/src/features/pieces/Piece.ts
export class Piece {
  constructor(
    public readonly id: number,
    public title: string,
    public restored: boolean,
    public readonly createdAt: Date,
  ) {}

  static describe(title: string): Piece {
    if (!title || title.trim().length === 0) {
      throw new Error("A piece must have a name");
    }
    return new Piece(0, title.trim(), false, new Date());
  }

  toggle(): void {
    this.restored = !this.restored;
  }

  rename(title: string): void {
    if (!title || title.trim().length === 0) {
      throw new Error("A piece must have a name");
    }
    this.title = title.trim();
  }
}

export type PieceRow = {
  id: number;
  title: string;
  restored: number;
  created_at: string;
};
